problem_builder/mixins.py

```diff
@@ -185,21 +185,38 @@ def studio_view(self, _context=None):
 
 class StudentViewUserStateMixin(object):
     """
-    Mixin to provide student_view_user_state view
+    Mixin to provide student_view_user_state view.
+
+    To prevent unnecessary overloading of the build_user_state_data method,
+    you may specify the USER_STATE_FIELDS and USER_STATE_TRANSFORMS attributes
+    to customise build_user_state_data and student_view_user_state output.
+
+    USER_STATE_FIELDS: field names to include in build_user_state_data
+
+    USER_STATE_TRANSFORMS: a dict of field_names: function to run with
+        the xblock instance as the only parameter.
     """
     NESTED_BLOCKS_KEY = "components"
     INCLUDE_SCOPES = (Scope.user_state, Scope.user_info, Scope.preferences)
+    USER_STATE_FIELDS = ['num_attempts', 'completed', 'student_results', ]
+    USER_STATE_TRANSFORMS = {}
 
     def build_user_state_data(self, context=None):
         """
-        Returns a JSON representation of the student data of this XBlock,
+        Returns a dictionary of the student data of this XBlock,
         retrievable from the Course Block API.
         """
+
         result = {}
         for _, field in self.fields.iteritems():
-            if field.scope in self.INCLUDE_SCOPES:
+            # Only insert fields if their scopes and field names match
+            if field.scope in self.INCLUDE_SCOPES and field.name in self.USER_STATE_FIELDS:
                 result[field.name] = field.read_from(self)
 
+        # Add fields according to transforms dict
+        for key, func in self.USER_STATE_TRANSFORMS.iteritems():
+            result[key] = func(self)
```

Comment:
@bradmerlin nit: Overwriting values obtained from field is slightly suboptimal and might be a source of WTFs in the future. I'd suggest to merge it together with original loop:

```
for _, field in self.fields.iteritems():
    if scope in INCLUDE_SCOPES and name in INCLUDE_FIELDS:
        field_reader = USER_STATE_TRANSFORMS.get(field.name, field.read_from)
        result[field.name] = field_reader(self)
```

Also, (and I know naming is hard, so no pressure :)), but those `USER_STATE_TRANSFORMS` is misleading as those functions semantically are not transformers, but readers:

* Transformer takes one values and transforms it to some other - i.e. it takes field value and maps it to something else
* Reader takes an object that contains some value and returns that value - i.e. it takes XBlock and return a value (not necessarily read from a field)

So it might make sense to rename the collection as well.